api/v1/attributes.proto

```diff
@@ -0,0 +1,69 @@
+// Copyright 2016 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+package istio.mixer.v1;
+
+option go_package="istio.io/mixer/api/v1;mixerpb";
+
+// Contains a set of attributes related to an event.
+message Attributes {
```

Comment:
I don't actually believe the premise underlying this optimization for several reasons:

- We have an existence proof with ESF & Chemist within GCP (for the non-Googlers, ESF is the logical equivalent to the Istio proxy and Chemist is the logical equivalent to the Istio mixer). In the ESF/Chemist world, we use long-form labels with every request. Check calls typically have 2-3 labels, Report calls have ~10 labels. Labels have pretty long scoped names, and often long values too. There is no delta functionality provided so labels are sent anew for every single request. This demonstrates that the overall model is definitely usable at scale in its current form.

- Communication with the various backends we interact with for logging and monitoring use long-form labels. Even if we optimize the proxy->mixer protocol, the mixer->backend communication will still be on the slow fully-expanded plan.

- With delta facts, only a minimal set of facts need to be transmitted from proxy to mixer. I can imagine that perhaps the 'fact context' model that Mandar suggested could further improve the efficiency of the deltas.

- Proxy-side aggregation, batching, and caching means that QPS between the proxy and the mixer is typically considerably lower than the inbound QPS to the proxy. So the communication overhead is amortized over many incoming calls to the proxy.

So I definitely would rather we stay on the current plan until we demonstrate there is value in something different.